### PR TITLE
znc: run as user znc & use procd

### PR DESCRIPTION
**Pull request by kdarbyshirebryant**
Maintainer: @KanjiMonster 
Compile tested: ar71xx, Archer C7 v2, LEDE r1519
Run tested: ar71xx, Archer C7 v2, LEDE r1519,  New user & group 'znc' created.  znc runs as new user under procd control.

Description:

Create & run znc as a specific user rather than nobody.  Converted to
use procd, removing dependencies on znc's 'droproot' module & 'su'

Signed-off-by: Kevin Darbyshire-Bryant <kevin@darbyshire-bryant.me.uk>

### CODE REVIEW

File: net/znc/files/znc.init [hunk 1]
```diff
@@ -156,45 +155,27 @@ add_user() {
...
-		kill $(cat "$PID_FILE")
-	else
-		killall znc
-	fi
+	procd_open_instance
+	procd_set_param file /etc/config/znc
+	[ "$EXTERNAL_CONFIG" -eq 1 ] && procd_set_param file "${ZNC_CONFIG}/configs/znc.conf"
+	procd_set_param command /usr/bin/znc
+	procd_append_param command -f -d$ZNC_CONFIG_PATH
+	procd_set_param user ${RUNAS_USER:-nobody}
```
Comment (karlp) [inline net/znc/files/znc.init]: Should this still have a fallback?  the chown earlier has no fallbacks?

Comment (kdarbyshirebryant) [inline net/znc/files/znc.init]: Thinking about it, probably not.  It was an ill thought out idea to cope with a missing user ID.  I've removed it.

File: net/znc/files/znc.init [hunk 2]
```diff
@@ -12,7 +15,6 @@ DISABLED=
 
 RUNAS_USER=
 RUNAS_GROUP=
-RUNAS_SHELL=
```
Comment (KanjiMonster) [inline net/znc/files/znc.init]: Please don't drop features silently. This was the place where I stopped my try at the conversion, because I couldn't at a first glance how to convert this to procd (or how it is used) so I chased other targets.

@Xotic750 can you maybe describe where this feature is needed, and how one can test whether it works?

File: net/znc/files/znc.init [hunk 3]
```diff
@@ -156,45 +155,30 @@ add_user() {
...
-
-start() {
+start_service() {
 	config_load znc
 	config_foreach znc_global znc
 
-	if [ "$DISABLED" -eq 1 ]; then
-		return 0
-	fi
+	[ "$DISABLED" -eq 0 ] || return 0
```
Comment (KanjiMonster) [inline net/znc/files/znc.init]: What's the reason for this change?

File: net/znc/Makefile [hunk 4]
```diff
@@ -31,6 +31,7 @@ define Package/znc/default
   CATEGORY:=Network
   TITLE:=ZNC
   URL:=http://en.znc.in/
+  USERID:=znc=65532:znc=65532
```
Comment (KanjiMonster) [inline net/znc/Makefile]: Maybe use 6667 for user/group? (the IRC default port). That would match what some other daemons chose for their ids.

File: net/znc/Makefile [hunk 5]
```diff
@@ -62,12 +63,11 @@ define Package/znc/install
 	$(INSTALL_DIR) $(1)/etc/config
 	$(INSTALL_DATA) ./files/znc.conf $(1)/etc/config/znc
 	$(INSTALL_DIR) $(1)/usr/lib/znc/
-	$(INSTALL_BIN) $(PKG_BUILD_DIR)/modules/droproot.so $(1)/usr/lib/znc/
 endef
 
 
 ZNC_MODULES :=
-ZNC_MODULE_TARGETS := droproot.so
```
Comment (KanjiMonster) [inline net/znc/Makefile]: If you remove the droproot module, you can also remove the patch reverting the removal in the sources.

File: net/znc/files/znc.init (hunk 3 above)
Comment (kdarbyshirebryant) [inline net/znc/files/znc.init]: I felt a non-zero value would indicate disabled.  In previous code if someone set the value to anything other than 1, then it mean enabled.

File: net/znc/files/znc.init (hunk 2 above)
Comment (kdarbyshirebryant) [inline net/znc/files/znc.init]: My understanding of this is that 'su' is used to switch user.  If that user doesn't have a valid shell then the 'su' will fail.
I tested znc with the 'shell' module and znc user shell set to /bin/false and all was well....although why on earth you'd give shell access via irc!!!!

Don't forget that this change also removes a dependency on 'su' being installed.

File: net/znc/Makefile (hunk 4 above)
Comment (kdarbyshirebryant) [inline net/znc/Makefile]: Personally I'd prefer to keep unprivileged 'system/daemon' users out of the way and close to other similar user accounts...e.g. nobody.  Whilst I can see your logic in '6667', I don't like it 'cos it's not low enough to scream 'system ID' at me and not high enough to be 'out of the way'  - but if that's the difference between accept/reject I really don't care.

File: net/znc/Makefile (hunk 5 above)
Comment (kdarbyshirebryant) [inline net/znc/Makefile]: Done

File: net/znc/files/znc.init (hunk 2 above)
Comment (Xotic750) [inline net/znc/files/znc.init]: That's correct, if `su` is used with a user without a shell then it would fail. As `nobody` was being used, and had no shell, then this was necessary for how it was written.

File: net/znc/Makefile (hunk 4 above)
Comment (yousong) [inline net/znc/Makefile]: Just in case it may help, there is no need to determine the id at packaging time.  It can be simply `USERID=znc:znc` and opkg will decide a free uid/gid pair at `postinst` stage.

**Comment by diizzyy** [comment]
I really think this is a change in the right direction, apart from hardcoding a userid/groupid which I think is a really bad idea unless we have a proper list somewhere (which we don't?) yousong's suggestion is a more sensible approach. "Randomly" setting userid/groupid is doomed to fail at some point unless we start to maintain a list in the package repo rootdir or perhaps a wiki page on Github?

Is that a sane tradeoff?

@kdarbyshirebryant @KanjiMonster @yousong

**Comment by kdarbyshirebryant** [comment]
Do whatever you like.  I no longer care.
